packages/@aws-cdk/cdk/lib/app.ts

```diff
@@ -34,73 +33,49 @@ export class App extends Root {
   /**
    * Runs the program. Output is written to output directory as specified in the request.
    */
-  public run(): void {
+  public run(): ISynthesisSession {
+    // this app has already been executed, no-op for you
+    if (this._session) {
+      return this._session;
+    }
+
     const outdir = process.env[cxapi.OUTDIR_ENV];
-    if (!outdir) {
-      process.stderr.write(`ERROR: The environment variable "${cxapi.OUTDIR_ENV}" is not defined\n`);
-      process.stderr.write('AWS CDK Toolkit (>= 0.11.0) is required in order to interact with this program.\n');
-      process.exit(1);
-      return;
+    if (outdir) {
+      this._session = new SynthesisSession({ outdir });
+    } else {
+      this._session = new InMemorySynthesisSession();
     }
 
-    const result: cxapi.SynthesizeResponse = {
-      version: cxapi.PROTO_RESPONSE_VERSION,
-      stacks: this.synthesizeStacks(Object.keys(this.stacks)),
-      runtime: this.collectRuntimeInformation()
-    };
+    // the three holy phases of synthesis: validate, prepare and synthesize
+    const errors = this.node.validateTree();
+    if (errors.length > 0) {
+      const errorList = errors.map(e => `[${e.source.node.path}] ${e.message}`).join('\n  ');
+      throw new Error(`Validation failed with the following errors:\n  ${errorList}`);
+    }
 
-    const outfile = path.join(outdir, cxapi.OUTFILE_NAME);
-    fs.writeFileSync(outfile, JSON.stringify(result, undefined, 2));
+    this.node.prepareTree();
+    this.node.synthesizeTree(this.run());
+
+    this._session.finalize(); // lock session - cannot emit more artifacts
+
+    return this._session;
   }
 
   /**
    * Synthesize and validate a single stack
    * @param stackName The name of the stack to synthesize
    */
   public synthesizeStack(stackName: string): cxapi.SynthesizedStack {
```

Comment:
Shouldn't we just get rid of this function and put it on `ISynthesisSession`? That seems like 200% better.

Call it `getStack()` or something.

Reply:
`ISynthesisSession` doesn't know about "stacks" and "templates". It just knows about files that were emitted from this app. This is a convenience method (and also backwards compatible) for users/tests that wish to obtain the result of a stack's synthesis and examine it.

I'll raise an issue to revisit this API (soon!) and finalize the right approach for tests/etc.

Reply:
https://github.com/awslabs/aws-cdk/issues/1891